[CollectionView] Fixes for ObservableCollection usage on Android and iOS

rmarinho opened the pull request:
### Description of Change ###

On Android and iOS navigating from a page that had a ObservableCollection as source was causing a crash because the iNotifyPropertyChanged event wasn't unsubscribed. Implemented Disposable on IItemsViewSource, (this was also in the TODO list from @hartez ). 

### Issues Resolved ### 
<!-- Please use the format "fixes #xxxx" for each issue this PR addresses -->


### API Changes ###


Changed:
 - internal interface IItemsViewSource - > internal interface IItemsViewSource : IDisposable
Added 


### Platforms Affected ### 
<!-- Please list all platforms affected by these changes -->

- Core/XAML (all platforms)
- Android

### Behavioral/Visual Changes ###

Should add items to the collection view , and allow to navigate back and to another gallery

### Before/After Screenshots ### 


### Testing Procedure ###

Visit ObservableCodeCollectionViewGallery with empty source.

### PR Checklist ###

- [ ] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [x] Rebased on top of the target branch at time of PR
- [ ] Changes adhere to coding standard

rmarinho commented on Xamarin.Forms.Platform.Android/CollectionView/ItemsViewRenderer.cs:
```diff
@@ -295,8 +295,8 @@ void TearDownOldElement(ItemsView oldElement)
 
 			if (adapter != null)
 			{
-				adapter.Dispose();
```
We need to un hood our adapter before dispose it.

rmarinho commented on Xamarin.Forms.Platform.iOS/CollectionView/ItemsViewLayout.cs:
```diff
@@ -120,13 +122,18 @@ public override bool ShouldInvalidateLayoutForBoundsChange(CGRect newBounds)
 				return;
 			}
 
+			if (!(GetPrototype() is ItemsViewCell prototype))
```
At this time we could not have a prototype because we don't have cells.

rmarinho commented on Xamarin.Forms.Platform.iOS/CollectionView/ObservableItemsSource.cs:
```diff
@@ -10,17 +10,21 @@ internal class ObservableItemsSource : IItemsViewSource
 
 		void CollectionChanged(object sender, NotifyCollectionChangedEventArgs args)
 		{
+			ShouldDetermineCellSize();
```
We are adding our first cell, the collection was empty and now we will need to DetermineCellSize

rmarinho commented on Xamarin.Forms.Platform.iOS/CollectionView/CollectionViewController.cs:
```diff
@@ -42,7 +42,14 @@ public override UICollectionViewCell GetCell(UICollectionView collectionView, NS
 		{
-			return _itemsSource.Count;
+			var count = _itemsSource.Count;
+			if (_layout.IsDeterminingCellSize && count == 1)
```
I would like a better idea for this I don't have one right now. 
So basically when you set EstimateItemSize on `DetermineCellSize` It will queue a call to GetItemsCount, since the source already has 1 item, but we called this method before adding a NativeCell to the UICollectionView, so to return the correct count we need to return 0.

mattleibow replied:
I may not be understanding this, but what happens if there are 2 items in the source? Why do we check specifically for `1` item? What if there is no items? Does this mean the size might be wrong? If the size is going to be incorrect, can we postpone the measuring until we have an item?

mattleibow commented on Xamarin.Forms.Platform.iOS/CollectionView/CollectionViewController.cs:
```diff
@@ -42,7 +42,14 @@ public override UICollectionViewCell GetCell(UICollectionView collectionView, NS
+			var count = _itemsSource.Count;
+			if (_layout.IsDeterminingCellSize && count == 1)
+			{
+				_layout.FinishedDeterminingCellSize();
```
There appears to be a disconnect here and the layout. It is `FinishedDeterminingCellSize` here, but `FinishDetermineCellSize` in the layout. Maybe this?

```suggestion
				_layout.FinishDeterminingCellSize();
```

mattleibow commented on Xamarin.Forms.Platform.iOS/CollectionView/ItemsViewLayout.cs:
```diff
@@ -155,7 +157,10 @@ protected void DetermineCellSize()
+		}
 
+		internal void FinishDetermineCellSize()
+		{
```
Here we have split the size calculation into 2 methods, but we never seem to call this. There is a possible connection and type on `FinishedDeterminingCellSize` in `CollectionViewController.GetItemsCount`, but this appears to be for just 1 type. How are these two methods supposed to be used?

If there are 2 methods that need to be called, do we have a way to test that both are bing called? For me, the split gets scary as we now have forced a second method that can be missed. And by looking at the current code, it will only ever run on iOS when there is a SINGLE item in the source.

rmarinho replied:
Fixed the call, it should be always called, but you are correct that a better way will be preferred. Need to see what @hartez thinks about it.

mattleibow commented on Xamarin.Forms.Platform.iOS/CollectionView/ObservableItemsSource.cs:
```diff
@@ -10,17 +10,21 @@ internal class ObservableItemsSource : IItemsViewSource
 		public ObservableItemsSource(IEnumerable itemSource, UICollectionView collectionView)
 		{
 			_collectionView = collectionView;
 			_itemsSource = (IList)itemSource;
```
Is this safe? I just noticed that we just cast. It might be better to change the constructor parameter to `IList` - this way we move the responsibility to the caller.

mattleibow commented on Xamarin.Forms.Platform.iOS/CollectionView/ObservableItemsSource.cs:
```diff
@@ -10,17 +10,21 @@ internal class ObservableItemsSource : IItemsViewSource
 			_itemsSource = (IList)itemSource;
 
 			((INotifyCollectionChanged)itemSource).CollectionChanged += CollectionChanged;
+			_isEmpty = !_itemsSource.GetEnumerator().MoveNext();
```
If this really is an `IList`, cant we rather do `Count`? This should be better.

rmarinho replied:
Done

hartez commented on Xamarin.Forms.Platform.Android/CollectionView/ItemsViewAdapter.cs:
```diff
@@ -16,9 +17,10 @@ namespace Xamarin.Forms.Platform.Android
-		readonly Context _context;
-		readonly Func<IVisualElementRenderer, Context, AView> _createView;
-		readonly IItemsViewSource _itemsSource;
+		Context _context;
```
See https://github.com/xamarin/Xamarin.Forms/pull/4569#pullrequestreview-179602894

hartez commented on Xamarin.Forms.Platform.iOS/CollectionView/ObservableItemsSource.cs:
```diff
@@ -11,10 +11,10 @@ internal class ObservableItemsSource : IItemsViewSource
 		readonly IList _itemsSource;
 
-		public ObservableItemsSource(IEnumerable itemSource, UICollectionView collectionView)
+		public ObservableItemsSource(IList itemSource, UICollectionView collectionView)
```
Why was this changed to IList?

mattleibow replied:
This was changed to `IList` because this source only understands lists, not any arb collection type. For example:
`public object this[int index]` and `public int Count`. What was happening was the constructor was just doing a straight cast, but this could cause issues in the future. Since this was just for observable collections, they are `IList`. In addition the factory has it defined as: `case IList _ when itemsSource is INotifyCollectionChanged:` - basically `IList` again.

I foresee evil where Bob comes along and says: "ooh, look. I can wrap my queue with this and magic will happen." and the only magic is a major 💥

hartez replied:
Well, it's `internal`, so Bob would have to slip it by us, but okay. I'm cool with changing it to `IList`, but for consistency it should also be changed in the Android version.

hartez commented on Xamarin.Forms.Platform.Android/CollectionView/ObservableItemsSource.cs:
```diff
@@ -23,6 +22,28 @@ public ObservableItemsSource(IEnumerable itemSource, RecyclerView.Adapter adapte
+				{
+					((INotifyCollectionChanged)_itemsSource).CollectionChanged -= CollectionChanged;
+					_itemsSource = null;
+					_adapter = null;
```
No point in setting `_adapter` and `_itemsSource` to `null` here, this instance is going away anyway.

hartez commented on Xamarin.Forms.Platform.Android/CollectionView/ObservableItemsSource.cs:
```diff
@@ -7,9 +7,8 @@ namespace Xamarin.Forms.Platform.Android
-		// TODO hartez 2018/07/30 14:40:11 We may need to implement IDisposable to make sure this all gets cleaned up	
-		readonly RecyclerView.Adapter _adapter;
-		readonly IList _itemsSource;
+		RecyclerView.Adapter _adapter;
```
Since we don't need to `null` these out, they don't actually change and we can leave these as `readonly`.

hartez commented on Xamarin.Forms.Platform.Android/CollectionView/ListSource.cs:
```diff
@@ -21,5 +23,24 @@ public ListSource(IEnumerable enumerable)
 			}
 		}
+
+		public void Dispose()
```
Since we aren't actually inheriting from this anywhere right now, we could just marked it `sealed` and then we don't have to implement the `protected virtual void Dispose(bool disposing)` pattern at all.

hartez commented on Xamarin.Forms.Platform.Android/CollectionView/ItemsViewAdapter.cs:
```diff
@@ -66,6 +65,22 @@ public override RecyclerView.ViewHolder OnCreateViewHolder(ViewGroup parent, int
+				if (disposing)
+				{
+					_itemsSource?.Dispose();
+					_itemsSource = null;
```
Setting `_itemsSource` and `_createView` to null doesn't do anything here, these can be removed.

hartez commented on Xamarin.Forms.Platform.Android/CollectionView/ItemsViewAdapter.cs:
```diff
@@ -1,23 +1,22 @@
 		protected readonly ItemsView ItemsView;
-		readonly Func<IVisualElementRenderer, Context, AView> _createView;
-		readonly IItemsViewSource _itemsSource;
+		Func<IVisualElementRenderer, Context, AView> _createView;
```
Since these don't need to be modified (see comments below), the `readonly` can be reinstated.

hartez commented on Xamarin.Forms.Platform.Android/CollectionView/ItemsViewAdapter.cs:
```diff
@@ -66,6 +65,22 @@ public override RecyclerView.ViewHolder OnCreateViewHolder(ViewGroup parent, int
 
+		protected override void Dispose(bool disposing)
+		{
+			if (!_disposed)
```
If this is already disposed, it shouldn't call `base.Dispose()` again.

hartez commented on Xamarin.Forms.Platform.Android/CollectionView/EmptySource.cs:
```diff
@@ -7,5 +7,25 @@ internal class EmptySource : IItemsViewSource
+
+		bool _disposed;
+
+		protected virtual void Dispose(bool disposing)
```
We can just mark this as `sealed` and then we don't have to implement the `protected virtual void Dispose(bool disposing)` pattern.

rmarinho commented:
build

samhouts commented:
@rmarinho Please remove the do-not-merge when you've updated the PR. Thanks!

rmarinho commented:
I rebased and reworked a little bit this issue, I also saw the same issue using the sample I added with the timer "Add Items with timer to Empty Collection" on iOS and added the fix.
